src/qcodes/parameters/infer.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import TYPE_CHECKING, ClassVar
+
+from qcodes.instrument import Instrument, InstrumentBase, InstrumentModule
+from qcodes.instrument.parameter import DelegateParameter, Parameter
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+DOES_NOT_EXIST = "Does not exist"
+
+
+class InferError(AttributeError): ...
+
+
+class InferAttrs:
+    """Holds a global set of attribute name that will be inferred"""
+
+    _known_attrs: ClassVar[set[str]] = set()
+
+    @classmethod
+    def add_attrs(cls, attrs: str | Iterable[str]) -> None:
+        if isinstance(attrs, str):
+            attrs = (attrs,)
+        cls._known_attrs.update(set(attrs))
+
+    @classmethod
+    def known_attrs(cls) -> tuple[str, ...]:
+        return tuple(cls._known_attrs)
+
+    @classmethod
+    def discard_attr(cls, attr: str) -> None:
+        cls._known_attrs.discard(attr)
+
+    @classmethod
+    def clear_attrs(cls) -> None:
+        cls._known_attrs = set()
+
+
+def get_root_parameter(
+    param: Parameter,
+    alt_source_attrs: Sequence[str] | None = None,
+) -> Parameter:
+    """Return the root parameter in a chain of DelegateParameters or other linking Parameters"""
+    alt_source_attrs_set = _merge_user_and_class_attrs(alt_source_attrs)
+
+    if isinstance(param, DelegateParameter):
+        if param.source is None:
+            raise InferError(f"Parameter {param} is not attached to a source")
+        return get_root_parameter(param.source)
+
+    for alt_source_attr in alt_source_attrs_set:
+        alt_source = getattr(param, alt_source_attr, DOES_NOT_EXIST)
+        if alt_source is None:
```

Comment:
Do I understand correctly that the DOES_NOT_EXIST case is simply ignored? E.g. we don’t tell the user that al_source_attr doesn’t exist or something else?

Reply:
Correct. DOES_NOT_EXIST distinguishes between two cases
1. The linking attribute is there, but no linking parameter has been added. This should return None, resulting in stopping the chain or an error
2. The linking attribute didn't exist in the first place. Which means this *isn't* a linking parameter of a type recognized by the inference.